cli-plugins/manager/hooks.go

```diff
@@ -0,0 +1,140 @@
+package manager
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/docker/cli/cli/command"
+	"github.com/morikuni/aec"
+	"github.com/spf13/cobra"
+)
+
+// HookPluginData is the type representing the information
+// that plugins declaring support for hooks get passed when
+// being invoked following a CLI command execution
+type HookPluginData struct {
+	RootCmd string
+}
+
+// HookMessage represents a plugin hook response. Plugins
+// declaring support for CLI hooks need to print a json
+// representation of this type when their hook subcommand
+// is invoked.
+type HookMessage struct {
+	// TODO: should we have an enum to typify the kind of hook
+	// (such as `Next Steps`, `Warning`, etc.)?
+	Template string
+}
+
+// TemplateReplaceSubcommandName returns a hook template string
+// that will be replaced by the CLI subcommand being executed
+//
+// Example:
+//
+//	h := HookMessage{
+//	  Template: `you ran the subcommand: ` + TemplateReplaceSubcommandName(),
+//	}
+//
+// when being executed after the command:
+// `docker run --name "my-container" alpine`
+// will result in the message:
+// `you ran the subcommand: run`
+func TemplateReplaceSubcommandName() string {
+	return hookTemplateCommandName
+}
+
+// TemplateReplaceFlagValue returns a hook template string
+// that will be replaced by the flags value.
+//
+// Example:
+//
+//	h := HookMessage{
+//	  Template: `you ran a container named: ` + TemplateReplaceFlagValue("name"),
+//	}
+//
+// when being executed after the command:
+// `docker run --name "my-container" alpine`
+// will result in the message:
+// `you ran a container named: my-container`
+func TemplateReplaceFlagValue(flag string) string {
+	return fmt.Sprintf(hookTemplateFlagValue, flag)
+}
+
+// RunPluginHooks calls the hook subcommand for all present
+// CLI plugins that declare support for hooks in their metadata
+// and parses/prints their responses
+func RunPluginHooks(dockerCli command.Cli, cmd *cobra.Command, args []string) error {
+	plugins, err := ListPlugins(dockerCli, cmd)
+	if err != nil {
+		return err
+	}
+	subCommand, _, err := cmd.Find(args)
+	if err != nil {
+		return err
+	}
+
+	hookMessages := make([]string, 0, len(plugins))
+	for _, p := range plugins {
+		p := p
+		if !p.supportsHookCommand(subCommand.Name()) {
```

Comment:
Perhaps we should store `Name()` in a variable; we're calling it in a loop here, and I noticed that it's not a plain "getter", but doing `strings.Index()` and generating the name on-the-fly.

Might be "micro-optimisation", but could be low-hanging fruit as well 😅